File: dsa_code/sorting/Bucket Sort.py
arr = [2, 1, 2, 0, 0, 2]


def bucket_sort(arr):
    counts_num = len(set(arr))
    counts = [0] * counts_num
    # print(counts)
    for n in arr:
        counts[n] += 1
    # print(counts)
    i = 0  # for insert the number in the arr's index
    for n in range(len(counts)):
        for j in range(counts[n]):
            arr[i] = n
            i += 1
    return arr


res = bucket_sort(arr)
print(res)

File: dsa_code_java/sorting/BucketSort.java
package dsa_code_java.sorting;

import java.util.HashSet;
import java.util.Set;
import java.util.Arrays;

public class BucketSort {
    public static void main(String[] args) {
        int[] arr = new int[] { 2, 1, 2, 0, 0, 2 };
        int[] res = bucketSort(arr);
        System.out.println("Sorted Array: " + Arrays.toString(res));
    }

    public static int[] bucketSort(int[] arr) {
        Set<Integer> uniqueElements = new HashSet<>();
        for (int num : arr) {
            uniqueElements.add(num);
        }
        int countsNum = uniqueElements.size();
        int[] counts = new int[countsNum + 1];

        for (int n : arr) {
            counts[n]++;
        }

        int i = 0; // index for arr
        for (int n = 0; n < counts.length; n++) {
            for (int j = 0; j < counts[n]; j++) {
                arr[i] = n;
                i++;
            }
        }

        return arr;
    }
}

// check the python logic is very simple
